### ISSUE TITLE
Integrate awareframework

### ISSUE THREAD

**Issue by rkdarst** [CxAalto/koota-server#27]
- QR code made in application/views/researcher_study.php 
  - defined line 42-53
  - around line 90 ti is presented to user
  - also something in application/controllers/researcher.php, not sure if it is used.
- Has a link to a config json file
  - application/controllers/webservice.php, line 102, starts making the JSON returned by the config url.
- Must use a MQTT server for messaging.

**Comment by rkdarst** [comment]
In order for clients to connect to MQTT server:
- under aware study, set study URL, then click activate.
- This results in a GET of /public/server.crt, which is used to verify the server for sending data.
- Code in https://github.com/denzilferreira/aware-client/blob/master/aware-core/src/main/java/com/aware/utils/SSLManager.java#L47

The things it is subscribed to are:

```
1465063193: Received SUBSCRIBE from -1651525856
1465063193:     UID/broadcasts (QoS 2)
1465063193:     UID/esm (QoS 2)
1465063193:     UID/configuration (QoS 2)
1465063193:     UID/# (QoS 2)
```

**Comment by rkdarst** [comment]
**QR codes:**

summary: the QR code has _only_ a URL, and it routes to index.php/webservice/index/$study_id/$api_key.

When I make my own QR code, it does not ever scan it and does not seem to react to it.  Does this following QR code cause any reaction in the iOS app?  (ping: @arashbm)  Once scanned, it won't do anything, but I can check server logs for activity.  https://koota.cs.aalto.fi/static/test-qr2.png

https://github.com/denzilferreira/aware-server/blob/master/application/views/researcher_study.php#L42

```
$join_url = base_url() . 'index.php/webservice/index/' . $study_data["id"] . '/' . $study_data["api_key"];
$qr_img_properties = array(
          'src' => 'https://chart.googleapis.com/chart?chs=300x300&cht=qr&chl=' . $join_url . '&choe=UTF-8',
          'id' => 'qr_img',
          'alt' => 'Study QRcode',
          'title' => 'Study QRcode',
          'class' => 'qrcode',
          'width' => '300',
          'height' => '300'
);
```

**Comment by rkdarst** [comment]
**After QR code is scanned:**
It will POST to the study_url (from the QR code, or from your manual entering of the study URL) with this body:
`b'study_check=1&device_id=UID'`
The below is done using manually entering the study_url.

It will first GET DOMAIN/public/server.crt and use that as the MQTT server certificate _and maybe_ as a https certificate.  It tends to crash around this point, but if you restart then it continues.

"POST /aware/register/de883177c61a056e HTTP/1.1" 500 328 32
"POST /aware/register/de883177c61a056e/aware_device/create_table HTTP/1.1" 500 912 32
"POST /aware/register/de883177c61a056e/applications_foreground/create_table HTTP/1.1" 500 6
"POST /aware/register/de883177c61a056e/applications_history/create_table HTTP/1.1" 500 748
"POST /aware/register/de883177c61a056e/applications_notifications/create_table HTTP/1.1" 50
"POST /aware/register/de883177c61a056e/applications_crashes/create_table HTTP/1.1" 500 775
"POST /aware/register/de883177c61a056e/aware_device/create_table HTTP/1.1" 500 912 32

```
kdata.aware: Aware request: /aware_device/create_table
kdata.aware: <QueryDict: {}>
kdata.aware: b"fields=_id%20integer%20primary%20key%20autoincrement%2Ctimestamp%20real%20default%200%2Cdevice_id%20text%20default%20''%2Cboard%20text%20default%20''%2Cbrand%20text%20default%20''%2Cdevice%20text%20default%20''%2Cbuild_id%20text%20default%20''%2Chardware%20text%20default%20''%2Cmanufacturer%20text%20default%20''%2Cmodel%20text%20default%20''%2Cproduct%20text%20default%20''%2Cserial%20text%20default%20''%2Crelease%20text%20default%20''%2Crelease_type%20text%20default%20''%2Csdk%20integer%20default%200%2Clabel%20text%20default%20''%2CUNIQUE%20(timestamp%2Cdevice_id)&device_id=UID"
kdata.aware: Aware request: /aware_device/latest
kdata.aware: <QueryDict: {}>
kdata.aware: b'device_id=a7890c06-c7e5-4b3f-a20e-58757ada0fe8'
```

**Comment by rkdarst** [comment]
**How to communicate with device**
- https://github.com/denzilferreira/aware-server/blob/master/application/controllers/webservice.php#L8
  - This is the function that returns the basic study data (from JSON).  aware/register should return this.
- https://github.com/denzilferreira/aware-server/blob/master/application/controllers/webservice.php#L149
  - This makes MQTT messages.  Why does it end in `echo json_message` ?  Does that return to user?
    - I think that researcher.js uses it, it POSTs a form to that function, and gets data backe.  Then the JS sends it to be published (I haven't checked the second part yet)
  - 
- This sends things via MQTT.  
  - https://github.com/denzilferreira/aware-server/blob/master/application/controllers/webservice.php#L372

**Comment by rkdarst** [comment]
When posting to webservice/index/study_id/api_key/$OPERATION
- $OPERATION will call various things on $aware_model
- https://github.com/denzilferreira/aware-server/blob/master/application/controllers/webservice.php#L125
- index/ returns:
  - JSON with study config + mqtt config
  - study config from researcher model: https://github.com/denzilferreira/aware-server/blob/master/application/models/researcher_model.php#L316  get_study_configuration, table studies_configuration
  - Directly from the `config` column in the `studies_configurations` in the database table
  - This is set in controllers/researcher.php
  - Then it has a ['sensors'] which gets the MQTT server stuff.

Aside:
- https://github.com/denzilferreira/aware-server/blob/master/application/models/researcher_model.php#L577  sensor configurations (get_sensor_configurations)

**Comment by rkdarst** [comment]
These also appear in the new app:

```
 "POST /index.php/aware/register/de883177c61a056e/mqtt_messages/create_table HTTP/1.1" 200 5
"POST /index.php/aware/register/de883177c61a056e/mqtt_messages/latest HTTP/1.1" 200 345 31
"POST /index.php/aware/register/de883177c61a056e/mqtt_subscriptions/create_table HTTP/1.1" 
"POST /index.php/aware/register/de883177c61a056e/mqtt_subscriptions/latest HTTP/1.1" 200 35
```

**Comment by rkdarst** [comment]
client config pipeline:
- Get config: https://github.com/denzilferreira/aware-server/blob/master/application/controllers/webservice.php#L85  
  - `$this->Researcher_model->get_study_configuration($study_id);`
- https://github.com/denzilferreira/aware-server/blob/master/application/controllers/webservice.php#L110
  - **Update config with per-device variables** in the index function.
- https://github.com/denzilferreira/aware-server/blob/master/application/models/researcher_model.php#L317
  - get_stufy_config function
  - config column from studies_configurations database.  This is JSON data
- https://github.com/denzilferreira/aware-server/blob/master/application/controllers/researcher.php#L383
  - update_study_configuration function (this calls update_study_config).  
  - config text: $config = str_replace(',', ';', $this->input->post('config', true));
  - is this JSON or form data?
- https://github.com/denzilferreira/aware-server/blob/master/application/models/researcher_model.php#L82
  - researcher_model.insert_new_study
  - Does not seem to make config here, inserts config = `[ ]`
  - line 328 (get_study_configuration) shows that this is how the study config comes out: return $result_array[0]['config'];  (probably selecting "config" from first row.)
- https://github.com/denzilferreira/aware-server/blob/master/application/controllers/researcher.php#L622
  - update_study_sensors
  - basically goes through json data and recreates it in a validated fashion.
- https://github.com/denzilferreira/aware-server/blob/master/application/models/researcher_model.php#L577
  - get_sensor_configurations
  - returns all the valid settings
  - https://github.com/denzilferreira/aware-server/blob/master/aware_dashboard.sql#L298
    - List of all sensor settings!

```
config format:
config['sensors'] = { }    # (php ordered map)
config['sensors'][0]  = {setting: name, value: x  }
config['plugins'] = { }
config['plugins'][0] = { plugin:name, settings = xxx }
xxx = { 0:{value=x, (?? name=x),},  1:{value=:, ...} }   ????
config['plugins'][0] 
```
